### PR TITLE
Change default for weatherEndpoint according to API 3.0

### LINKED ISSUE #3574: Weather Module not loading since MM update
Hi all,

i did the MM Update few days ago (Notifikation in the bottom bar) via git pull. Until the Update everything worked fine, but since the Update was done, the Weather Module is not loading any more. All other modules are fine after the update. I dont see anything in the logs. Would be great if someone could help me figuring out whats going on.

System: Raspi Zero 2W with PoE LAN Hat
Node: v22.9.0
OS: Debian 12 Bookworm latest update/upgrade
Electron: v32.1.2
MagicMirror Version: 2.29.0

Weather Module API is OpenWeather Map
API Key is valid and checked.

Config-Part of Weather:
```
		{
			module: "weather",
			position: "bottom_right",
			header: "Wetter Vorhersage",
			config: {
				weatherProvider: "openweathermap",
				type: "forecast",
				maxNumberOfDays: 5,
				locationID: "2831948", //ID from http://bulk.openweathermap.org/sample/city.list.json.gz; unzip the gz file and find your city
				apiKey: "*****CENSORED*******",
				weatherEndpoint: "/forecast",
				tableClass: 'medium',
				colored: 'true',
				fade: true,
				fadePoint: 0.5,
				
			}
```

Log output if restarting MM:

> 0|mm  | [2024-10-02 12:37:42.326] [LOG]   Starting MagicMirror: v2.29.0 
> 0|mm  | [2024-10-02 12:37:42.427] [LOG]   Loading config ...
> 0|mm  | [2024-10-02 12:37:42.436] [LOG]   config template file not exists, no envsubst
> 0|mm  | [2024-10-02 12:37:43.863] [INFO]  Checking config file /home/pi/MagicMirror/config/config.js ...
> 0|mm  | [2024-10-02 12:37:44.087] [INFO]  Your configuration file doesn't contain syntax errors :)
> 0|mm  | [2024-10-02 12:37:44.090] [INFO]  Checking modules structure configuration ...
> 0|mm  | [2024-10-02 12:37:44.205] [INFO]  Your modules structure configuration doesn't contain errors :)
> 0|mm  | [2024-10-02 12:37:44.212] [LOG]   Loading module helpers ...
> 0|mm  | [2024-10-02 12:37:44.218] [LOG]   No helper found for module: alert.
> 0|mm  | [2024-10-02 12:37:44.240] [LOG]   Initializing new module helper ...
> 0|mm  | [2024-10-02 12:37:44.242] [LOG]   Module helper loaded: updatenotification
> 0|mm  | [2024-10-02 12:37:44.245] [LOG]   No helper found for module: clock.
> 0|mm  | [2024-10-02 12:37:44.701] [LOG]   Initializing new module helper ...
> 0|mm  | [2024-10-02 12:37:44.706] [LOG]   Module helper loaded: calendar
> 0|mm  | [2024-10-02 12:37:44.717] [LOG]   No helper found for module: MMM-AutoDimmer.
> 0|mm  | [2024-10-02 12:37:44.723] [LOG]   No helper found for module: weather.
> 0|mm  | [2024-10-02 12:37:44.741] [LOG]   Initializing new module helper ...
> 0|mm  | [2024-10-02 12:37:44.742] [LOG]   Module helper loaded: netatmo
> 0|mm  | [2024-10-02 12:37:44.743] [LOG]   All module helpers loaded.
> 0|mm  | [2024-10-02 12:37:44.764] [LOG]   Starting server on port 8080 ...
> 0|mm  | [2024-10-02 12:37:56.138] [LOG]   Server started ...
> 0|mm  | [2024-10-02 12:37:56.163] [LOG]   Connecting socket for: updatenotification
> 0|mm  | [2024-10-02 12:37:56.165] [LOG]   Starting module helper: updatenotification
> 0|mm  | [2024-10-02 12:37:56.166] [LOG]   Connecting socket for: calendar
> 0|mm  | [2024-10-02 12:37:56.168] [LOG]   Starting node helper for: calendar
> 0|mm  | [2024-10-02 12:37:56.169] [LOG]   Connecting socket for: netatmo
> 0|mm  | [2024-10-02 12:37:56.170] [LOG]   Netatmo helper started ...
> 0|mm  | [2024-10-02 12:37:56.172] [LOG]   Sockets connected & modules started ...
> 0|mm  | [2024-10-02 12:37:58.210] [LOG]   Launching application.
> 0|mm  | [2024-10-02 12:38:09.296] [INFO]  System information:
> 0|mm  | ### SYSTEM:   manufacturer: Raspberry Pi Foundation; model: Raspberry Pi Zero 2 W Rev 1.0; virtual: false
> 0|mm  | ### OS:       platform: linux; distro: Raspbian GNU/Linux; release: 12; arch: arm; kernel: 6.6.51+rpt-rpi-v7
> 0|mm  | ### VERSIONS: electron: 29.1.6; used node: 20.9.0; installed node: 22.9.0; npm: 10.8.3; pm2: 5.3.1
> 0|mm  | ### OTHER:    timeZone: Europe/Berlin; ELECTRON_ENABLE_GPU: undefined
> 0|mm  | [2024-10-02 12:38:18.041] [ERROR] Netatmo: clientId not set in config.
> 0|mm  | [2024-10-02 12:38:18.116] [LOG]   Create new calendarfetcher for url: http://p71-caldav.icloud.com/published/2/MjA0MDUwMTIxMjA.....*CENSORED* - Interval: 604800000
> 0|mm  | [2024-10-02 12:38:18.786] [LOG]   Netatmo: using token.json file
> 0|mm  | [2024-10-02 12:38:18.794] [LOG]   Netatmo: Initialized
> 0|mm  | [2024-10-02 12:38:18.847] [INFO]  updatenotification: Updater Class Loaded!
> 0|mm  | [2024-10-02 12:38:18.848] [INFO]  updatenotification: Checking PM2 using...
> 0|mm  | [2024-10-02 12:38:21.265] [INFO]  Checking git for module: MMM-AutoDimmer
> 0|mm  | [2024-10-02 12:38:21.642] [INFO]  Checking git for module: netatmo
> 0|mm  | [2024-10-02 12:38:21.819] [INFO]  Checking git for module: MagicMirror
> 0|mm  | [2024-10-02 12:38:21.972] [INFO]  updatenotification: [PM2] You are not using pm2
> 0|mm  | [2024-10-02 12:38:22.050] [LOG]   Netatmo: Authenticated
> 0|mm  | [2024-10-02 12:38:22.053] [LOG]   Netatmo: token.json was written successfully
> 0|mm  | [2024-10-02 12:38:22.109] [LOG]   Netatmo: New Token Expire Wednesday, October 2, 2024 3:38 PM
> 0|mm  | [2024-10-02 12:38:32.845] [ERROR] Moment Timezone has no data for GMT+0200. See http://momentjs.com/timezone/docs/#/data-loading/.
> 0|mm  | [2024-10-02 12:38:32.846] [ERROR] Moment Timezone has no data for GMT+0200. See http://momentjs.com/timezone/docs/#/data-loading/.
> 0|mm  | [2024-10-02 12:38:32.847] [ERROR] Moment Timezone has no data for GMT+0200. See http://momentjs.com/timezone/docs/#/data-loading/.
> 0|mm  | [2024-10-02 12:38:33.062] [INFO]  Calendar-Fetcher: Broadcasting 40 events from http://p71-caldav.icloud.com/published/2/MjA0MDUwMTIxMjA0MD........*CENSORED*.

I dont even see the weather module loading in the log-File? But to be honest, i didnt compeared if it was listed before the update in the log... Unfortunatly i didnt stored any older log files.
The Display shows endless "loading". Can someone have a look on this or maybe guide me into deeper analysis?

image

### PR DESCRIPTION
**Pull request by HeikoGr**
since API 3.0 is default, weatherEndpoint should be set to "/onecall"
Fixes #3574

ATTENTION: since lat / lon  defaults to 0 / 0, the weather plugins works after this patch, but shows the weather from https://de.wikipedia.org/wiki/Null_Island if lat / lon is not manually set.

### CODE REVIEW

**Comment by sdetweil** [comment]
as weather is all in browser we could log a message there about default lat/lon

**Comment by khassel** [comment]
@HeikoGr the tests are failing because the `/` before `onecall`, must be `weatherEndpoint: "onecall",`

I fixed the merge conflicts in changelog so you should fetch your `master` locally before working on this ... (can do that too if you want)

**Comment by sdetweil** [comment]
funny, on the module side its "/onecall"

**Comment by khassel** [comment]
> funny, on the module side its "/onecall"

upsi, I tested this without and test works

> as weather is all in browser we could log a message there about default lat/lon

we should think about more changes, e.g. 
- removing `openweathermap` as default provider (leave empty or choose another)
- update code so `weatherEndpoint` works with and without leading slash
- ...

**Comment by khassel** [comment]
will fail again, now e2e tests, will check why ...

**Comment by HeikoGr** [comment]
ok, i will wait for your response

**Comment by khassel** [comment]
question: is `onecall` the new default for all weather types, for current weather, for hourly weather and for forecast?

**Comment by sdetweil** [comment]
openweather says /onecall is only

**Comment by HeikoGr** [comment]
Except for pro customers. This is why i didn‘t remove the other endpoint functions

**Comment by sdetweil** [comment]
but our users are mostly not pro. so does onecall
pull back everything and you can sort it out

**Comment by khassel** [comment]
so now I know what is happening. First thing `/onecall` is correct (with `/`).

Problem is, that if no `weatherEndpoint` is specified we have this function in the code:

```js
	/**
	 * Overrides method for setting config to check if endpoint is correct for hourly
	 * @param {object} config The configuration object
	 */
	setConfig (config) {
		this.config = config;
		if (!this.config.weatherEndpoint) {
			switch (this.config.type) {
				case "hourly":
					this.config.weatherEndpoint = "/onecall";
					break;
				case "daily":
				case "forecast":
					this.config.weatherEndpoint = "/forecast";
					break;
				case "current":
					this.config.weatherEndpoint = "/weather";
					break;
				default:
					Log.error("weatherEndpoint not configured and could not resolve it based on type");
			}
		}
	},
```

So with a new default `"/onecall"` instead of `""` some tests fail because their configs have no `weatherEndpoint` and no `weatherProvider` and the above function is not called anymore.

To discuss:
- all weather tests and (mock data?) are based on provider openweathermap. Is this correct @rejas ?
- if the tests are bound to the above provider we should update their configs with `weatherEndpoint` and `weatherProvider`
- do we need the function `setConfig` if we set a default `weatherEndpoint`? I think not, pro users can still set `weatherEndpoint` in the config

**Comment by sdetweil** [comment]
we should test without the endpoint set too
should not be required

pick a provider, it should work

**Comment by rejas** [comment]
> all weather tests and (mock data?) are based on provider openweathermap. Is this correct @rejas ?

correct, we only test against openweathermap, so the mock data is specific to that provider too.

File: modules/default/weather/providers/openweathermap.js [hunk 1]
```diff
@@ -17,10 +17,13 @@ WeatherProvider.register("openweathermap", {
 	defaults: {
 		apiVersion: "3.0",
 		apiBase: "https://api.openweathermap.org/data/",
-		weatherEndpoint: "", // can be "onecall", "forecast" or "weather" (for current)
+		// weatherEndpoint is "onecall" since API 3.0
+		// "onecall", "forecast" or "weather" only for pro customers
+		weatherEndpoint: "onecall",
```
Comment (pedrolamas) [inline modules/default/weather/providers/openweathermap.js:22]: Shouldn't this be "/onecall" ?

```suggestion
		weatherEndpoint: "/onecall",
```

Comment (khassel) [reply modules/default/weather/providers/openweathermap.js:22]: yes, see my comment above - but we have to change more things to get the tests working

**Comment by khassel** [comment]
@HeikoGr I have a fix for the tests now, is it o.k. to push this on your branch?

**Comment by HeikoGr** [comment]
absolutely

**Review by rejas — APPROVED** [review]

**Pull request merged**